File: src/client/acontext-py/src/acontext/messages.py
"""
Support for constructing session messages.
"""

from collections.abc import Mapping, MutableMapping, Sequence
from dataclasses import dataclass
from typing import Any, Literal

@dataclass(slots=True)
class MessagePart:
    """
    Represents a single message part for ``/session/{id}/messages``.

    Args:
        type: One of ``text``, ``image``, ``audio``, ``video``, ``file``, ``tool-call``,
            ``tool-result`` or ``data``.
        text: Optional textual payload for ``text`` parts.
        meta: Optional metadata dictionary accepted by the API.
        file_field: Optional field name to use in the multipart body. When omitted the
            client will auto-generate deterministic field names.
    """

    type: str
    text: str | None = None
    meta: Mapping[str, Any] | None = None
    file_field: str | None = None

    @classmethod
    def text_part(cls, text: str, *, meta: Mapping[str, Any] | None = None) -> "MessagePart":
        return cls(type="text", text=text, meta=meta)
    
    @classmethod
    def file_field_part(cls, file_field: str, *, meta: Mapping[str, Any] | None = None) -> "MessagePart":
        return cls(type="file", file_field=file_field, meta=meta)

@dataclass(slots=True)
class AcontextMessage:
    """
    Represents an Acontext-format message payload.
    """

    role: Literal["user", "assistant", "system"]
    parts: list[MessagePart]
    meta: MutableMapping[str, Any] | None = None


def build_acontext_message(
    *,
    role: Literal["user", "assistant", "system"],
    parts: Sequence[MessagePart | str | Mapping[str, Any]],
    meta: Mapping[str, Any] | None = None,
) -> AcontextMessage:
    """
    Construct an Acontext-format message blob and associated multipart files.
    """
    if role not in {"user", "assistant", "system"}:
        raise ValueError("role must be one of {'user', 'assistant', 'system'}")

    normalized_parts = [normalize_message_part(part) for part in parts]

    message = AcontextMessage(
        role=role,
        parts=normalized_parts,
        meta=dict(meta) if meta is not None else None,
    )
    return message


def normalize_message_part(part: MessagePart | str | Mapping[str, Any]) -> MessagePart:
    if isinstance(part, MessagePart):
        return part
    if isinstance(part, str):
        return MessagePart(type="text", text=part)
    if isinstance(part, Mapping):
        if "type" not in part:
            raise ValueError("mapping message parts must include a 'type'")
        return MessagePart(
            type=str(part["type"]),
            text=part.get("text"),
            meta=part.get("meta"),
            file_field=part.get("file_field"),
        )
    raise TypeError("unsupported message part type")

File: src/client/acontext-ts/src/messages.ts
/**
 * Support for constructing session messages.
 */

import { z } from 'zod';

export const MessagePartSchema = z.object({
  type: z.string(),
  text: z.string().nullable().optional(),
  meta: z.record(z.string(), z.unknown()).nullable().optional(),
  file_field: z.string().nullable().optional(),
});

export type MessagePartInput = z.infer<typeof MessagePartSchema>;

export class MessagePart {
  type: string;
  text?: string | null;
  meta?: Record<string, unknown> | null;
  file_field?: string | null;

  constructor(data: MessagePartInput) {
    this.type = data.type;
    this.text = data.text ?? null;
    this.meta = data.meta ?? null;
    this.file_field = data.file_field ?? null;
  }

  static textPart(
    text: string,
    options?: { meta?: Record<string, unknown> | null }
  ): MessagePart {
    return new MessagePart({
      type: 'text',
      text,
      meta: options?.meta ?? null,
    });
  }

  static fileFieldPart(
    fileField: string,
    options?: { meta?: Record<string, unknown> | null }
  ): MessagePart {
    return new MessagePart({
      type: 'file',
      file_field: fileField,
      meta: options?.meta ?? null,
    });
  }

  toJSON(): MessagePartInput {
    return {
      type: this.type,
      text: this.text ?? null,
      meta: this.meta ?? null,
      file_field: this.file_field ?? null,
    };
  }
}

export const AcontextMessageSchema = z.object({
  role: z.enum(['user', 'assistant', 'system']),
  parts: z.array(MessagePartSchema),
  meta: z.record(z.string(), z.unknown()).nullable().optional(),
});

export type AcontextMessageInput = z.infer<typeof AcontextMessageSchema>;

export class AcontextMessage {
  role: 'user' | 'assistant' | 'system';
  parts: MessagePart[];
  meta?: Record<string, unknown> | null;

  constructor(data: AcontextMessageInput) {
    this.role = data.role;
    this.parts = data.parts.map((p) =>
      p instanceof MessagePart ? p : new MessagePart(p)
    );
    this.meta = data.meta ?? null;
  }

  toJSON(): AcontextMessageInput {
    return {
      role: this.role,
      parts: this.parts.map((p) => p.toJSON()),
      meta: this.meta ?? null,
    };
  }
}

export function buildAcontextMessage(options: {
  role: 'user' | 'assistant' | 'system';
  parts: (MessagePart | string | MessagePartInput)[];
  meta?: Record<string, unknown> | null;
}): AcontextMessage {
  if (!['user', 'assistant', 'system'].includes(options.role)) {
    throw new Error("role must be one of {'user', 'assistant', 'system'}");
  }

  const normalizedParts = options.parts.map((part) => {
    if (part instanceof MessagePart) {
      return part;
    }
    if (typeof part === 'string') {
      return MessagePart.textPart(part);
    }
    if (typeof part === 'object' && part !== null) {
      if (!('type' in part)) {
        throw new Error("mapping message parts must include a 'type'");
      }
      return new MessagePart(part as MessagePartInput);
    }
    throw new TypeError('unsupported message part type');
  });

  return new AcontextMessage({
    role: options.role,
    parts: normalizedParts,
    meta: options.meta ?? null,
  });
}

